packages/grid/_modules_/grid/utils/rowTreeUtils.ts

```diff
@@ -0,0 +1,117 @@
+import type { GridRowTreeNodeConfig, GridRowId, GridRowTreeConfig } from '../models';
+import type {
+  GridRowGroupingResult,
+  GridRowGroupParams,
+} from '../hooks/core/rowGroupsPerProcessing';
+
+type GridNodeNameToIdTree = {
+  [nodeName: string]: { id: GridRowId; children: GridNodeNameToIdTree };
+};
+
+interface GenerateRowTreeParams extends GridRowGroupParams {
+  rows: { id: GridRowId; path: string[] }[];
+  defaultGroupingExpansionDepth: number;
+}
+
+interface TempRowTreeNode extends Omit<GridRowTreeNodeConfig, 'children'> {
+  children?: Record<GridRowId, GridRowId>;
+}
+
+/**
+ * Transform a list of rows into a tree structure where each row references its parent and children.
+ * Add the auto generated rows to `ids` and `idRowsLookup`.
+ */
+export const buildRowTree = (params: GenerateRowTreeParams): GridRowGroupingResult => {
+  // During the build, we store the children as a Record to avoid linear complexity when checking if a children is already defined.
+  const tempTree: Record<GridRowId, TempRowTreeNode> = {};
+  let treeDepth = 1;
+
+  const ids = [...params.ids];
+  const idRowsLookup = { ...params.idRowsLookup };
+
+  const nodeNameToIdTree: GridNodeNameToIdTree = {};
+
+  for (let i = 0; i < params.rows.length; i += 1) {
+    const row = params.rows[i];
+    let nodeNameToIdSubTree = nodeNameToIdTree;
+    let parentNode: TempRowTreeNode | null = null;
+
+    for (let depth = 0; depth < row.path.length; depth += 1) {
+      const nodeName = row.path[depth];
+      let nodeId: GridRowId;
+
+      const expanded =
+        params.defaultGroupingExpansionDepth === -1 || params.defaultGroupingExpansionDepth > depth;
+
+      let nodeNameConfig = nodeNameToIdSubTree[nodeName];
+
+      if (!nodeNameConfig) {
+        nodeId =
+          depth === row.path.length - 1
+            ? row.id
+            : `auto-generated-row-${row.path.slice(0, depth + 1).join('-')}`;
+
+        nodeNameConfig = { id: nodeId, children: {} };
+        nodeNameToIdSubTree[nodeName] = nodeNameConfig;
+      } else {
+        nodeId = nodeNameConfig.id;
+      }
+      nodeNameToIdSubTree = nodeNameConfig.children;
+
+      if (depth < row.path.length - 1) {
+        let node = tempTree[nodeId] ?? null;
+        if (!node) {
```

Comment:
Could you comment the code of this function? I spent a good time to understand that here it's filling the missing values.

Example: https://github.com/dowjones/react-dropdown-tree-select/blob/26371757d8f861f534c4b4f23167d16730c091ac/src/tree-manager/flatten-tree.js#L9

We never did that but this helper function could have unit tests to make easier to see the input and output.

Reply:
I agree that this kind of function would be worth testing.

Reply:
I added a few tests